File: java/src/main/java/com/leetcode/ThreeEightFour.java
package com.leetcode;

import java.util.Random;

/**
 * @author: John
 * @date: 2020-03-25 16:37
 * @description: 384
 **/
public class ThreeEightFour {
    private int[] array;
    private int[] original;
    Random rand = new Random();

    private int randRange(int min, int max) {
        return rand.nextInt(max - min) + min;
    }
    private void swapAt(int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }

    public ThreeEightFour(int[] nums) {
       array = nums;
       original = nums.clone();
    }

    public int[] reset() {
        array = original;
        original = original.clone();
        return original;
    }

    public int[] shuffle() {
        for (int i = 0; i < array.length; i++) {
            swapAt(i, randRange(i, array.length));
        }
        return array;
    }

}

File: leetcodecpp/leetcode/ThreeEightFour.cpp
//
// Created by John on 2020/3/25.
//
#include "vector"

using namespace std;

class Solution{
public:
    Solution(vector<int> nums) {
        Elements = nums;
    }
    // Resets the Array to its original configuration and return it.
    vector<int> reset() {
        return Elements;
    }

    // Return a random shuffling of the array.
    vector<int> shuffle() {
        vector<int> vShuffle = Elements;
        for (int i = 1; i < vShuffle.size(); ++i) {
            int r = rand() % (i + 1);
            if (r != i) {
                swap(vShuffle[r], vShuffle[i]);
            }
        }
        return vShuffle;
    }

private:
    vector<int> Elements;
};

//int main() {
//    vector<int> nums;
//    nums.push_back(1);
//    nums.push_back(2);
//    nums.push_back(3);
//    Solution * solution = new Solution(nums);
//    solution->shuffle();
//}